File: code-tiny/tools/project_topology/registry.py
"""Descriptor, primary-analyzer, and framework coverage registries."""

from __future__ import annotations

import fnmatch
from dataclasses import dataclass
from pathlib import PurePosixPath
from types import MappingProxyType
from typing import Callable, Iterable, Mapping, Optional, Tuple

from .models import DescriptorParseOutput, DescriptorRole, ParseDepth
from .parsers import (
    parse_ant,
    parse_android_manifest,
    parse_android_resource,
    parse_cmake,
    parse_gradle_build,
    parse_gradle_settings,
    parse_identity_manifest,
    parse_make,
    parse_maven,
    parse_protobuf,
)


ParserCallable = Callable[..., DescriptorParseOutput]


@dataclass(frozen=True)
class DescriptorSpec:
    name: str
    patterns: Tuple[str, ...]
    parser: ParserCallable
    role: DescriptorRole
    parse_depth: ParseDepth
    max_bytes: int = 2 * 1024 * 1024
    secret_bearing: bool = False
    generated: bool = False

    def matches(self, path: str) -> bool:
        normalized = path.replace("\\", "/")
        basename = PurePosixPath(normalized).name
        return any(
            fnmatch.fnmatch(normalized, pattern)
            or fnmatch.fnmatch(basename, pattern)
            for pattern in self.patterns
        )

    def parse(self, *, project_id: str, path: str, text: str) -> DescriptorParseOutput:
        kwargs = {"project_id": project_id, "path": path, "text": text}
        if self.parser is parse_identity_manifest:
            kwargs.update(
                {
                    "parser": self.name,
                    "role": self.role,
                    "parse_depth": self.parse_depth,
                    "secret_bearing": self.secret_bearing,
                    "generated": self.generated,
                }
            )
        return self.parser(**kwargs)


DESCRIPTOR_SPECS: Tuple[DescriptorSpec, ...] = (
    DescriptorSpec(
        "gradle_settings",
        ("settings.gradle", "settings.gradle.kts"),
        parse_gradle_settings,
        DescriptorRole.TOPOLOGY,
        ParseDepth.TOPOLOGY,
    ),
    DescriptorSpec(
        "gradle_build",
        ("build.gradle", "build.gradle.kts"),
        parse_gradle_build,
        DescriptorRole.DEPENDENCY,
        ParseDepth.DEPENDENCY,
    ),
    DescriptorSpec(
        "maven",
        ("pom.xml",),
        parse_maven,
        DescriptorRole.DEPENDENCY,
        ParseDepth.DEPENDENCY,
    ),
    DescriptorSpec(
        "ant",
        ("build.xml",),
        parse_ant,
        DescriptorRole.TOPOLOGY,
        ParseDepth.TOPOLOGY,
    ),
    DescriptorSpec(
        "cmake",
        ("CMakeLists.txt",),
        parse_cmake,
        DescriptorRole.TOPOLOGY,
        ParseDepth.DEPENDENCY,
    ),
    DescriptorSpec(
        "make",
        ("Makefile", "makefile", "GNUmakefile"),
        parse_make,
        DescriptorRole.TOPOLOGY,
        ParseDepth.DEPENDENCY,
    ),
    DescriptorSpec(
        "protobuf",
        ("*.proto",),
        parse_protobuf,
        DescriptorRole.INTERFACE,
        ParseDepth.SEMANTIC,
    ),
    DescriptorSpec(
        "android_manifest",
        ("AndroidManifest.xml",),
        parse_android_manifest,
        DescriptorRole.FRAMEWORK,
        ParseDepth.SEMANTIC,
    ),
    DescriptorSpec(
        "android_resource",
        ("*/res/*.xml", "*/res/**/*.xml"),
        parse_android_resource,
        DescriptorRole.RESOURCE,
        ParseDepth.SEMANTIC,
    ),
    DescriptorSpec(
        "go",
        ("go.mod", "go.work"),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "rust",
        ("Cargo.toml",),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "dart",
        ("pubspec.yaml",),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "swift",
        ("Package.swift", "project.pbxproj"),
        parse_identity_manifest,
        DescriptorRole.TOPOLOGY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "dotnet",
        ("*.csproj", "*.vbproj", "*.sln", "*.slnx"),
        parse_identity_manifest,
        DescriptorRole.TOPOLOGY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "python",
        ("pyproject.toml", "setup.cfg", "setup.py", "Pipfile"),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "javascript",
        ("package.json", "jsconfig.json", "tsconfig.json"),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "php",
        ("composer.json",),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "perl",
        ("cpanfile", "META.json", "META.yml", "dist.ini", "Makefile.PL", "Build.PL"),
        parse_identity_manifest,
        DescriptorRole.DEPENDENCY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "delphi",
        ("*.dproj", "*.groupproj", "*.dpk", "*.dpr"),
        parse_identity_manifest,
        DescriptorRole.TOPOLOGY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "visual_basic",
        ("*.vbp", "*.vbg", "*.vbw"),
        parse_identity_manifest,
        DescriptorRole.TOPOLOGY,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "database_migration",
        ("dbt_project.yml", "liquibase*.xml", "flyway*.conf"),
        parse_identity_manifest,
        DescriptorRole.DEPLOYMENT,
        ParseDepth.IDENTITY,
    ),
    DescriptorSpec(
        "framework_config",
        (
            "application*.properties",
            "application*.yml",
            "application*.yaml",
            "web.xml",
            "struts*.xml",
            "mybatis*.xml",
            "appsettings*.json",
        ),
        parse_identity_manifest,
        DescriptorRole.FRAMEWORK,
        ParseDepth.IDENTITY,
        secret_bearing=True,
    ),
    DescriptorSpec(
        "runtime_secret_keys",
        (".env", ".env.*", "secrets.json", "gradle.properties"),
        parse_identity_manifest,
        DescriptorRole.SECRET_BEARING,
        ParseDepth.IDENTITY,
        secret_bearing=True,
    ),
    DescriptorSpec(
        "generated_lock",
        (
            "*.lock",
            "package-lock.json",
            "yarn.lock",
            "pnpm-lock.yaml",
            "go.sum",
            "Package.resolved",
        ),
        parse_identity_manifest,
        DescriptorRole.GENERATED,
        ParseDepth.IDENTITY,
        generated=True,
        max_bytes=8 * 1024 * 1024,
    ),
)


@dataclass(frozen=True)
class CoverageEntry:
    name: str
    patterns: Tuple[str, ...]
    roles: Tuple[DescriptorRole, ...]
    parse_depth: ParseDepth
    fixture_id: str
    secret_policy: str = "redact-values"
    generated_policy: str = "lower-priority"


PRIMARY_SPECIAL_FILE_COVERAGE: Mapping[str, CoverageEntry] = MappingProxyType(
    {
        "android": CoverageEntry("android", ("AndroidManifest.xml", "settings.gradle*", "build.gradle*"), (DescriptorRole.IDENTITY, DescriptorRole.TOPOLOGY), ParseDepth.SEMANTIC, "project-topology/android"),
        "cobol": CoverageEntry("cobol", ("*.cpy", "*.jcl", "Makefile"), (DescriptorRole.INTERFACE, DescriptorRole.DEPLOYMENT), ParseDepth.IDENTITY, "parser-special-files/cobol"),
        "cplus": CoverageEntry("cplus", ("CMakeLists.txt", "Makefile", "compile_commands.json"), (DescriptorRole.TOPOLOGY, DescriptorRole.GENERATED), ParseDepth.DEPENDENCY, "project-topology/native"),
        "csharp": CoverageEntry("csharp", ("*.csproj", "*.sln", "appsettings*.json"), (DescriptorRole.TOPOLOGY, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "parser-special-files/dotnet"),
        "dart": CoverageEntry("dart", ("pubspec.yaml", "pubspec.lock"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/dart"),
        "delphi": CoverageEntry("delphi", ("*.dproj", "*.groupproj", "*.dfm"), (DescriptorRole.TOPOLOGY, DescriptorRole.RESOURCE), ParseDepth.IDENTITY, "parser-special-files/delphi"),
        "go": CoverageEntry("go", ("go.mod", "go.work", "go.sum"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/go"),
        "java": CoverageEntry("java", ("pom.xml", "settings.gradle*", "module-info.java"), (DescriptorRole.TOPOLOGY, DescriptorRole.INTERFACE), ParseDepth.DEPENDENCY, "project-topology/maven"),
        "js": CoverageEntry("js", ("package.json", "*lock*", "jsconfig.json"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/javascript"),
        "kotlin": CoverageEntry("kotlin", ("settings.gradle*", "build.gradle*", "pom.xml"), (DescriptorRole.TOPOLOGY, DescriptorRole.DEPENDENCY), ParseDepth.DEPENDENCY, "project-topology/gradle"),
        "php": CoverageEntry("php", ("composer.json", "composer.lock"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/php"),
        "perl": CoverageEntry("perl", ("cpanfile", "META.*", "dist.ini"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/perl"),
        "plsql": CoverageEntry("plsql", ("*.sql", "liquibase*.xml", "flyway*.conf"), (DescriptorRole.INTERFACE, DescriptorRole.DEPLOYMENT), ParseDepth.IDENTITY, "parser-special-files/plsql"),
        "python": CoverageEntry("python", ("pyproject.toml", "requirements*.txt", "*.lock"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/python"),
        "rust": CoverageEntry("rust", ("Cargo.toml", "Cargo.lock"), (DescriptorRole.IDENTITY, DescriptorRole.DEPENDENCY), ParseDepth.IDENTITY, "parser-special-files/rust"),
        "sql": CoverageEntry("sql", ("*.sql", "dbt_project.yml", "liquibase*.xml"), (DescriptorRole.INTERFACE, DescriptorRole.DEPLOYMENT), ParseDepth.IDENTITY, "parser-special-files/sql"),
        "swift": CoverageEntry("swift", ("Package.swift", "project.pbxproj", "Info.plist"), (DescriptorRole.TOPOLOGY, DescriptorRole.RESOURCE), ParseDepth.IDENTITY, "parser-special-files/swift"),
        "ts": CoverageEntry("ts", ("package.json", "tsconfig.json", "*lock*"), (DescriptorRole.IDENTITY, DescriptorRole.TOPOLOGY), ParseDepth.IDENTITY, "parser-special-files/typescript"),
        "vb6": CoverageEntry("vb6", ("*.vbp", "*.vbg", "*.frm"), (DescriptorRole.TOPOLOGY, DescriptorRole.RESOURCE), ParseDepth.IDENTITY, "parser-special-files/vb6"),
        "vba": CoverageEntry("vba", ("*.bas", "*.cls", "*.frm"), (DescriptorRole.INTERFACE, DescriptorRole.RESOURCE), ParseDepth.IDENTITY, "parser-special-files/vba"),
        "vbnet": CoverageEntry("vbnet", ("*.vbproj", "*.sln", "app.config"), (DescriptorRole.TOPOLOGY, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "parser-special-files/vbnet"),
        "vbscript": CoverageEntry("vbscript", ("*.vbs", "*.wsf", "web.config"), (DescriptorRole.IDENTITY, DescriptorRole.DEPLOYMENT), ParseDepth.IDENTITY, "parser-special-files/vbscript"),
    }
)


FRAMEWORK_CONTEXT_COVERAGE: Mapping[str, CoverageEntry] = MappingProxyType(
    {
        "aspnet_core": CoverageEntry("aspnet_core", ("*.csproj", "appsettings*.json", "Program.cs"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "framework-context/aspnet-core"),
        "aspnet_framework": CoverageEntry("aspnet_framework", ("*.csproj", "web.config", "Global.asax"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "framework-context/aspnet-framework"),
        "flutter": CoverageEntry("flutter", ("pubspec.yaml", "l10n.yaml"), (DescriptorRole.FRAMEWORK, DescriptorRole.RESOURCE), ParseDepth.IDENTITY, "framework-context/flutter"),
        "mybatis": CoverageEntry("mybatis", ("mybatis*.xml", "*Mapper.xml"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.SEMANTIC, "framework-context/mybatis"),
        "servlet_jsp": CoverageEntry("servlet_jsp", ("web.xml", "*.jsp", "*.tld"), (DescriptorRole.FRAMEWORK, DescriptorRole.RESOURCE), ParseDepth.SEMANTIC, "framework-context/servlet-jsp"),
        "spring": CoverageEntry("spring", ("application*.yml", "application*.properties", "pom.xml"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.SEMANTIC, "framework-context/spring"),
        "struts": CoverageEntry("struts", ("struts*.xml", "*-validation.xml"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.SEMANTIC, "framework-context/struts"),
        "fastapi_django": CoverageEntry("fastapi_django", ("manage.py", "settings.py", "urls.py"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "framework-context/python-web"),
        "express_js": CoverageEntry("express_js", ("package.json", "routes*.js"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "framework-context/express"),
        "laravel": CoverageEntry("laravel", ("composer.json", "artisan", "routes/*.php"), (DescriptorRole.FRAMEWORK, DescriptorRole.CONFIGURATION), ParseDepth.IDENTITY, "framework-context/laravel"),
        "database_sql": CoverageEntry("database_sql", ("*.sql", "dbt_project.yml", "liquibase*.xml"), (DescriptorRole.FRAMEWORK, DescriptorRole.DEPLOYMENT), ParseDepth.SEMANTIC, "framework-context/database-sql"),
        "database_plsql": CoverageEntry("database_plsql", ("*.sql", "flyway*.conf", "liquibase*.xml"), (DescriptorRole.FRAMEWORK, DescriptorRole.DEPLOYMENT), ParseDepth.SEMANTIC, "framework-context/database-plsql"),
    }
)


def descriptor_spec_for_path(path: str) -> Optional[DescriptorSpec]:
    return next((spec for spec in DESCRIPTOR_SPECS if spec.matches(path)), None)


def descriptor_candidates(paths: Iterable[str]) -> Tuple[str, ...]:
    return tuple(
        sorted(
            {
                path.replace("\\", "/")
                for path in paths
                if descriptor_spec_for_path(path) is not None
            }
        )
    )


__all__ = [
    "CoverageEntry",
    "DESCRIPTOR_SPECS",
    "DescriptorSpec",
    "FRAMEWORK_CONTEXT_COVERAGE",
    "PRIMARY_SPECIAL_FILE_COVERAGE",
    "descriptor_candidates",
    "descriptor_spec_for_path",
]

File: rust/crates/analyzer-topology/src/registry.rs
//! Port `tools/project_topology/registry.py` — descriptor spec registry với
//! fnmatch semantics (`fnmatch.fnmatch` POSIX: case-sensitive, `*` ăn cả '/').
//! Bảng specs giữ nguyên thứ tự — `descriptor_spec_for_path` trả spec đầu tiên
//! khớp, nên thứ tự là một phần của contract.

use crate::models::{descriptor_role, parse_depth};

#[derive(Debug, Clone, Copy)]
pub struct DescriptorSpec {
    pub name: &'static str,
    pub patterns: &'static [&'static str],
    pub parser: ParserKind,
    pub role: &'static str,
    pub parse_depth: &'static str,
    pub max_bytes: usize,
    pub secret_bearing: bool,
    pub generated: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ParserKind {
    GradleSettings,
    GradleBuild,
    Maven,
    Ant,
    Cmake,
    Make,
    Ini,
    Protobuf,
    AndroidManifest,
    AndroidResource,
    IdentityManifest,
}

impl DescriptorSpec {
    /// `DescriptorSpec.matches` — fnmatch(full_path) OR fnmatch(basename).
    pub fn matches(&self, path: &str) -> bool {
        let normalized = path.replace('\\', "/");
        let basename = normalized.rsplit('/').next().unwrap_or("");
        self.patterns
            .iter()
            .any(|pattern| fnmatch(pattern, &normalized) || fnmatch(pattern, basename))
    }
}

pub const DESCRIPTOR_SPECS: &[DescriptorSpec] = &[
    DescriptorSpec {
        name: "gradle_settings",
        patterns: &["settings.gradle", "settings.gradle.kts"],
        parser: ParserKind::GradleSettings,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::TOPOLOGY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "gradle_build",
        patterns: &["build.gradle", "build.gradle.kts"],
        parser: ParserKind::GradleBuild,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::DEPENDENCY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "maven",
        patterns: &["pom.xml"],
        parser: ParserKind::Maven,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::DEPENDENCY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "ant",
        patterns: &["build.xml"],
        parser: ParserKind::Ant,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::TOPOLOGY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "cmake",
        patterns: &["CMakeLists.txt"],
        parser: ParserKind::Cmake,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::DEPENDENCY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "make",
        patterns: &["Makefile", "makefile", "GNUmakefile"],
        parser: ParserKind::Make,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::DEPENDENCY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "ini",
        patterns: &["*.ini"],
        parser: ParserKind::Ini,
        role: descriptor_role::CONFIGURATION,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: true,
        generated: false,
    },
    DescriptorSpec {
        name: "protobuf",
        patterns: &["*.proto"],
        parser: ParserKind::Protobuf,
        role: descriptor_role::INTERFACE,
        parse_depth: parse_depth::SEMANTIC,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "android_manifest",
        patterns: &["AndroidManifest.xml"],
        parser: ParserKind::AndroidManifest,
        role: descriptor_role::FRAMEWORK,
        parse_depth: parse_depth::SEMANTIC,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "android_resource",
        patterns: &["*/res/*.xml", "*/res/**/*.xml"],
        parser: ParserKind::AndroidResource,
        role: descriptor_role::RESOURCE,
        parse_depth: parse_depth::SEMANTIC,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "go",
        patterns: &["go.mod", "go.work"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "rust",
        patterns: &["Cargo.toml"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "dart",
        patterns: &["pubspec.yaml"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "swift",
        patterns: &["Package.swift", "project.pbxproj"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "dotnet",
        patterns: &["*.csproj", "*.vbproj", "*.sln", "*.slnx"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "python",
        patterns: &["pyproject.toml", "setup.cfg", "setup.py", "Pipfile"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "javascript",
        patterns: &["package.json", "jsconfig.json", "tsconfig.json"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "php",
        patterns: &["composer.json"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "perl",
        patterns: &["cpanfile", "META.json", "META.yml", "dist.ini", "Makefile.PL", "Build.PL"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPENDENCY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "delphi",
        patterns: &["*.dproj", "*.groupproj", "*.dpk", "*.dpr"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "visual_basic",
        patterns: &["*.vbp", "*.vbg", "*.vbw"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::TOPOLOGY,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "database_migration",
        patterns: &["dbt_project.yml", "liquibase*.xml", "flyway*.conf"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::DEPLOYMENT,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: false,
        generated: false,
    },
    DescriptorSpec {
        name: "framework_config",
        patterns: &[
            "application*.properties",
            "application*.yml",
            "application*.yaml",
            "web.xml",
            "struts*.xml",
            "mybatis*.xml",
            "appsettings*.json",
        ],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::FRAMEWORK,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: true,
        generated: false,
    },
    DescriptorSpec {
        name: "runtime_secret_keys",
        patterns: &[".env", ".env.*", "secrets.json", "gradle.properties"],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::SECRET_BEARING,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 2 * 1024 * 1024,
        secret_bearing: true,
        generated: false,
    },
    DescriptorSpec {
        name: "generated_lock",
        patterns: &[
            "*.lock",
            "package-lock.json",
            "yarn.lock",
            "pnpm-lock.yaml",
            "go.sum",
            "Package.resolved",
        ],
        parser: ParserKind::IdentityManifest,
        role: descriptor_role::GENERATED,
        parse_depth: parse_depth::IDENTITY,
        max_bytes: 8 * 1024 * 1024,
        secret_bearing: false,
        generated: true,
    },
];

/// `descriptor_spec_for_path` — spec đầu tiên khớp theo thứ tự bảng.
pub fn descriptor_spec_for_path(path: &str) -> Option<&'static DescriptorSpec> {
    DESCRIPTOR_SPECS.iter().find(|spec| spec.matches(path))
}

/// `descriptor_candidates` — sorted unique paths có spec, backslash → slash.
pub fn descriptor_candidates(paths: impl IntoIterator<Item = String>) -> Vec<String> {
    let mut normalized: Vec<String> = paths
        .into_iter()
        .map(|path| path.replace('\\', "/"))
        .filter(|path| descriptor_spec_for_path(path).is_some())
        .collect();
    normalized.sort();
    normalized.dedup();
    normalized
}

/// `fnmatch.fnmatch` trên POSIX (case-sensitive) — `*` khớp mọi ký tự kể cả
/// '/', `?` một ký tự, `[seq]` character class (fnmatch.translate semantics).
pub fn fnmatch(pattern: &str, name: &str) -> bool {
    let regex_source = fnmatch_translate(pattern);
    match regex::Regex::new(&regex_source) {
        Ok(re) => re.is_match(name),
        Err(_) => false,
    }
}

/// `fnmatch.translate` — tương đương Python (không anchored dollar newline).
pub fn fnmatch_translate(pattern: &str) -> String {
    let mut out = String::from("(?s:");
    out.push_str("(?i:"); // không — POSIX fnmatch case-sensitive; bỏ (?i:
    out.clear();
    out.push_str("(?s:");
    let mut index = 0usize;
    let bytes = pattern.as_bytes();
    while index < bytes.len() {
        let ch = pattern[index..].chars().next().unwrap();
        let ch_len = ch.len_utf8();
        match ch {
            '*' => out.push_str(".*"),
            '?' => out.push('.'),
            '[' => {
                let mut j = index + 1;
                let mut negated = false;
                if j < bytes.len() && (bytes[j] == b'!' || bytes[j] == b'^') {
                    negated = true;
                    j += 1;
                }
                let mut class = String::from("[^");
                if !negated {
                    class = String::from("[");
                }
                let mut first = true;
                let mut closed = false;
                while j < bytes.len() {
                    let c = pattern[j..].chars().next().unwrap();
                    if c == ']' && !first {
                        closed = true;
                        j += c.len_utf8();
                        break;
                    }
                    first = false;
                    match c {
                        '\\' => {
                            class.push_str("\\\\");
                            j += 1;
                        }
                        ']' => {
                            class.push_str("\\]");
                            j += 1;
                        }
                        '^' => {
                            class.push_str("\\^");
                            j += 1;
                        }
                        _ => {
                            class.push(c);
                            j += c.len_utf8();
                        }
                    }
                }
                if closed {
                    class.push(']');
                    out.push_str(&class);
                    index = j;
                    continue;
                }
                out.push_str("\\[");
            }
            c => {
                if "\\.+()|^${}".contains(c) {
                    out.push('\\');
                }
                out.push(c);
            }
        }
        index += ch_len;
    }
    out.push_str(")$");
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn spec_matches_full_path_or_basename() {
        let spec = descriptor_spec_for_path("app/src/main/res/values/strings.xml").unwrap();
        assert_eq!(spec.name, "android_resource");
        let spec = descriptor_spec_for_path("app/src/main/AndroidManifest.xml").unwrap();
        assert_eq!(spec.name, "android_manifest");
        let spec = descriptor_spec_for_path("sub/settings.gradle.kts").unwrap();
        assert_eq!(spec.name, "gradle_settings");
        assert!(descriptor_spec_for_path("plain.xml").is_none());
    }

    #[test]
    fn fnmatch_star_crosses_slashes() {
        assert!(fnmatch("*/res/*.xml", "a/b/res/c.xml"));
        assert!(fnmatch("*/res/**/*.xml", "a/b/res/deep/c.xml"));
        assert!(!fnmatch("*/res/*.xml", "res/c.xml"));
        assert!(fnmatch("*.lock", "a/b/deep.lock"));
        assert!(fnmatch(".env.*", ".env.local"));
    }

    #[test]
    fn candidates_sorted_filtered() {
        let picked = descriptor_candidates([
            "b/build.gradle".to_string(),
            "a/pom.xml".to_string(),
            "random.txt".to_string(),
            "b/build.gradle".to_string(),
        ]);
        assert_eq!(picked, vec!["a/pom.xml".to_string(), "b/build.gradle".to_string()]);
    }
}
